tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala

```diff
@@ -144,17 +144,21 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext with LazyL
   private val statisticsLock: ReentrantLock = new ReentrantLock()
   // start a new thread to collect statistics
   private val statisticsThread = new Thread(() => {
-    while (state == Running()) {
+    var interrupted = false
+    while (state == Running() && !interrupted) {
       // Sleep for a while.
       // If we call printStatistics right away, we can easily run into a race condition with Z3 initializing.
       // This produces a core dump.
-      Thread.sleep(config.z3StatsSec * 1000)
       // make sure that the context is not being disposed right now. Otherwise, we can get a nice core dump.
       statisticsLock.lock()
```

Comment:
Grabbing the lock before you `Thread.sleep()` does not look right, it blocks out the other thread for a rather long time.
Why is it necessary?